routemaster/cron.py

```diff
@@ -0,0 +1,156 @@
+        if isinstance(state, Action):
+            scheduler.every().minute.do(
+                _process,
+                _retry_action,
+                state,
+            )
+
+        elif isinstance(state, Gate):
+            for trigger in state.triggers:
+                if isinstance(trigger, TimeTrigger):
+                    scheduler.every().day.at(
+                        f"{trigger.time.hour:02d}:{trigger.time.minute:02d}",
+                    ).do(
+                        _process,
+                        _trigger_gate,
+                        state,
+                    )
+                elif isinstance(trigger, IntervalTrigger):
+                    scheduler.every(
+                        trigger.interval.total_seconds(),
+                    ).seconds.do(
+                        _process,
+                        _trigger_gate,
+                        state,
+                    )
+                elif isinstance(trigger, MetadataTrigger):
+                    scheduler.every(60).seconds.do(
+                        _process,
+                        _retry_metadata_updates,
+                        state,
+                    )
+
+
+def configure_schedule(
+    app: App,
+    scheduler: schedule.Scheduler,
+    is_terminating: Callable[[], bool],
+) -> None:
+    """Set up all scheduled tasks that need running."""
+    for state_machine in app.config.state_machines.values():
+        _configure_schedule_for_state_machine(
+            app,
+            scheduler,
+            state_machine,
+            is_terminating,
+        )
+
+
+class CronThread(threading.Thread):  # pragma: no cover
+    """Background thread for running periodic jobs."""
+
+    def __init__(self, app: App) -> None:
+        self._terminating = False
+        self.app = app
+        self.scheduler = schedule.Scheduler()
+        super().__init__(name="cron", daemon=True)
+
+    def run(self) -> None:
+        """Run main scheduling loop."""
+        configure_schedule(self.app, self.scheduler, self.is_terminating)
+        while not self.is_terminating():
+            self.scheduler.run_pending()
+            time.sleep(1)
+
+    def stop(self) -> None:
+        """Set the stopping flag and wait for thread end."""
+        self._terminating = True
+        self.join()
+
+    def is_terminating(self) -> bool:
+        """Dynamically access whether we are terminating."""
+        return self._terminating
+
+
+class StopCronProcessing(Exception):
+    """Signal to exit the processing of a series of cron items."""
+    pass
+
+
+ProcessItemContextManager = Callable[
+    [],
+    contextlib.AbstractContextManager,
+]
+CronProcessor = Callable[[App, State, ProcessItemContextManager], None]
+
+
+def _process_cron_job(
+    is_terminating: Callable[[], bool],
+    fn: CronProcessor,
+    app: App,
+    state_machine: StateMachine,
+    state: State,
+) -> None:
+
+    @contextlib.contextmanager
+    def _inner() -> Iterator[None]:
```

Comment:
While this is a very neat mechanism, it's a bit surprising that a) it's a context manager, and b) a context manager throws an exception as part of its normal operations. It seems plausible that the exception could accidentally get caught by one of our blanket `except Exception`s (which we're using quite a lot of in this project).

Did you consider using something like a `threading.Event`, which would have clearer semantics?

Reply:
I think we're going to want to document whatever mechanism we go for somewhere so that it's clear what needs to be done to use it within other parts of the system.

Reply:
Having just realised that this manager also silences exceptions and spotted a bug with its usage in `routemaster.state_machines.actions.process_retries` (see comments there), I don't think that this is a good mechanism to use.
I think there's just a bit too much magic happening here for it to be safe in general use.